src/java.base/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java

```diff
@@ -219,24 +203,26 @@ private static void verifyUnboxBindings(Class<?> inType, List<Binding> bindings)
         }
     }
 
-    private static final Set<Binding.Tag> BOX_TAGS = EnumSet.of(
-        //VM_STORE,
-        VM_LOAD,
-        BUFFER_STORE,
-        //BUFFER_LOAD,
-        COPY_BUFFER,
-        ALLOC_BUFFER,
-        BOX_ADDRESS,
-        //UNBOX_ADDRESS,
-        DUP,
-        CAST
-    );
+    static boolean isUnbox(Binding binding) {
+        return switch (binding) {
+            case Binding.VMStore unused -> true;
```

Comment:
I wonder... here it might be better to capture the box/unbox nature of a binding in a virtual method in the binding class? E.g. isBox(), isUnbox() ?

Reply:
I prefer this approach, to be honest, since the logic for all the different binding operators is in one place. It gives an overview of which operators are expected in an unboxing recipe. Making them virtual methods would put quite a bit of visual distance between the different `true`/`false` values.

I've been through the Binding file quite a bit, and the amount of scrolling/searching needed to find a particular implementation is kind of annoying.

Reply:
To pull on that string some more. I think we should move the implementations of various Binding::verify methods here, or perhaps into a separate BindingVerifier class.

Ditto for the Binding::interpret implementations. They could be moved to BindingInterpreter. The Binding class would just be left as a simple bag of records + documentation for each operator.

The main motivation for this would be that, if you're looking at e.g. interpretation, most of the stuff in `Binding` is just noise that you have to filter through. I think it makes more sense to group these things into classes (for specialization/verification/interpretation), where all the code in a class is logically related.